### ISSUE TITLE
Fix watchdog

### ISSUE THREAD

**Issue by BarbourSmith** [MaslowCNC/Firmware#148]
I think it is possible to get in a position where the machine is ready to receive a new command, but Ground Control thinks it is not and so the UI becomes unresponsive. Clicking buttons won't do anything because Ground Control is waiting for a 'gready' command from the machine, and the machine thinks it has already sent the command so it's waiting for the next instruction.

I added a watchdog that sends the ready command every second, but that seems to have broken everything and caused very erratic behavior on the machine side. I need to figure out why and add a similar mechanism.

**Comment by johnbenweber** [comment]
Yes, I've found that the command sending is pretty finicky when you start changing stuff which is why I haven't implemented the gcode handling class I wrote. I think we need some more concrete communication instead of depending on kivy to do something when it detects variable changes. Depending on that functionality also makes the code flow reasonably hard to follow. I found that you can dispatch arbitrary events which you can bind to in other places. This seems to me to be more explicit and self documenting.

Back on topic, if you send multiple ready signals it makes sense that GroundControl won't be able to keep track of which command it was that finished triggering the ready message.

**Comment by BarbourSmith** [comment]
I agree 100%

**Comment by johnbenweber** [comment]
I think this may take a bit of work and testing, could you create a dedicated branch?

**Comment by BarbourSmith** [comment]
Absolutely. It's [here](https://github.com/MaslowCNC/GroundControl/tree/investigate-watchdog)

**Comment by BarbourSmith** [comment]
Added in #168 
@johnbenweber There is still more that can be done to improve the serial communication system, but I'm going to close this issue because the original problem of fixing the watchdog function is done.

**Comment by johnbenweber** [comment]
@BarbourSmith Looks like I've missed some fun while I was away. I'm working on getting some bigger motors to get my setup running. Everything is in place really, I just need more torque.

**Comment by BarbourSmith** [comment]
There's plenty of fun still to be had 😀 

Did you get the motors sorted? We just got our big shipment so we've got plenty of them now. Email me your address and I'm happy to send you some! ✈️
